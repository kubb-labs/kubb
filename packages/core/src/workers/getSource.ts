import { print } from '@kubb/parser'
import * as factory from '@kubb/parser/factory'

import isEqual from 'lodash.isequal'
import { orderBy } from 'natural-orderby'
import { z } from 'zod'

import { getRelativePath } from '../fs/read.ts'
import { searchAndReplace } from '../transformers/searchAndReplace.ts'
import { trimExtName } from '../transformers/trim.ts'
import { fileSchema } from './schemas.ts'

import type { KubbFile } from '../FileManager.ts'

function combineImports(imports: Array<KubbFile.Import>, exports: Array<KubbFile.Export>, source?: string): Array<KubbFile.Import> {
  const combinedImports = orderBy(imports, [(v) => !v.isTypeOnly], ['asc']).reduce((prev, curr) => {
    let name = Array.isArray(curr.name) ? [...new Set(curr.name)] : curr.name

    const hasImportInSource = (importName: string) => {
      if (!source) {
        return true
      }

      const checker = (name?: string) => name && !!source.includes(name)
      return checker(importName) || exports.some(({ name }) => (Array.isArray(name) ? name.some(checker) : checker(name)))
    }

    if (Array.isArray(name)) {
      name = name.filter((item) => typeof item === 'string' ? hasImportInSource(item) : hasImportInSource(item.propertyName))
    }

    const prevByPath = prev.findLast((imp) => imp.path === curr.path && imp.isTypeOnly === curr.isTypeOnly)
    const uniquePrev = prev.findLast((imp) => imp.path === curr.path && isEqual(imp.name, name) && imp.isTypeOnly === curr.isTypeOnly)
    const prevByPathNameAndIsTypeOnly = prev.findLast((imp) => imp.path === curr.path && isEqual(imp.name, name) && imp.isTypeOnly)

    if (prevByPathNameAndIsTypeOnly) {
      // we already have an export that has the same path but uses `isTypeOnly` (import type ...)
      return prev
    }

    if (uniquePrev || (Array.isArray(name) && !name.length)) {
      return prev
    }

    if (!prevByPath) {
      return [
        ...prev,
        {
          ...curr,
          name,
        },
      ]
    }

    if (prevByPath && Array.isArray(prevByPath.name) && Array.isArray(name) && prevByPath.isTypeOnly === curr.isTypeOnly) {
      prevByPath.name = [...new Set([...prevByPath.name, ...name])]

      return prev
    }

    if (!Array.isArray(name) && name && !hasImportInSource(name)) {
      return prev
    }

    return [...prev, curr]
  }, [] as Array<KubbFile.Import>)

  return orderBy(combinedImports, [(v) => !v.isTypeOnly], ['desc'])
}

function combineExports(exports: Array<KubbFile.Export>): Array<KubbFile.Export> {
  const combinedExports = orderBy(exports, [(v) => !v.isTypeOnly], ['asc']).reduce((prev, curr) => {
    const name = curr.name
    const prevByPath = prev.findLast((imp) => imp.path === curr.path)
    const prevByPathAndIsTypeOnly = prev.findLast((imp) => imp.path === curr.path && isEqual(imp.name, name) && imp.isTypeOnly)

    if (prevByPathAndIsTypeOnly) {
      // we already have an export that has the same path but uses `isTypeOnly` (export type ...)
      return prev
    }

    const uniquePrev = prev.findLast(
      (imp) => imp.path === curr.path && isEqual(imp.name, name) && imp.isTypeOnly === curr.isTypeOnly && imp.asAlias === curr.asAlias,
    )

    if (uniquePrev || (Array.isArray(name) && !name.length) || (prevByPath?.asAlias && !curr.asAlias)) {
      return prev
    }

    if (!prevByPath) {
      return [
        ...prev,
        {
          ...curr,
          name: Array.isArray(name) ? [...new Set(name)] : name,
        },
      ]
    }

    if (prevByPath && Array.isArray(prevByPath.name) && Array.isArray(curr.name) && prevByPath.isTypeOnly === curr.isTypeOnly) {
      prevByPath.name = [...new Set([...prevByPath.name, ...curr.name])]

      return prev
    }

    return [...prev, curr]
  }, [] as Array<KubbFile.Export>)

  return orderBy(combinedExports, [(v) => !v.isTypeOnly, (v) => v.asAlias], ['desc', 'desc'])
}

function getEnvSource(source: string, env: NodeJS.ProcessEnv | undefined): string {
  if (!env) {
    return source
  }

  const keys = Object.keys(env)

  if (!keys.length) {
    return source
  }

  return keys.reduce((prev, key: string) => {
    const environmentValue = env[key]
    const replaceBy = environmentValue ? `'${environmentValue.replaceAll('"', '')?.replaceAll("'", '')}'` : 'undefined'

    if (key.toUpperCase() !== key) {
      throw new TypeError(`Environment should be in upperCase for ${key}`)
    }

    if (typeof replaceBy === 'string') {
      prev = searchAndReplace({ text: prev.replaceAll(`process.env.${key}`, replaceBy), replaceBy, prefix: 'process.env', key })
      // removes `declare const ...`
      prev = searchAndReplace({ text: prev.replaceAll(new RegExp(`(declare const).*\n`, 'ig'), ''), replaceBy, key })
    }

    return prev
  }, source)
}

function getSource<TMeta extends KubbFile.FileMetaBase = KubbFile.FileMetaBase>(file: KubbFile.File<TMeta>): string {
  const exports = file.exports ? combineExports(file.exports) : []
  const imports = file.imports ? combineImports(file.imports, exports, file.source) : []

  const importNodes = imports.filter(item => {
    // isImportNotNeeded
    // trim extName
    return item.path !== trimExtName(file.path)
  }).map((item) => {
    return factory.createImportDeclaration({
      name: item.name,
      path: item.root ? getRelativePath(item.root, item.path) : item.path,
      isTypeOnly: item.isTypeOnly,
    })
  })
  const exportNodes = exports.map((item) =>
    factory.createExportDeclaration({
      name: item.name,
      path: item.path,
      isTypeOnly: item.isTypeOnly,
      asAlias: item.asAlias,
    })
  )

  return [print([...importNodes, ...exportNodes]), getEnvSource(file.source, file.env)].join('\n')
}

export const optionsSchema = z.object({
  file: fileSchema,
})

export default (options: z.infer<typeof optionsSchema>): string => {
  try {
    const result = optionsSchema.parse(options)

    return getSource(result.file)
  } catch (e) {
    console.log(e)
    return ''
  }
}
