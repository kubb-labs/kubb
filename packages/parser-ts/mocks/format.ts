import { format as prettierFormat } from 'prettier'
import pluginTypescript from 'prettier/plugins/typescript'

import type { Options } from 'prettier'

const formatOptions: Options = {
  tabWidth: 2,
  printWidth: 160,
  parser: 'typescript',
  singleQuote: true,
  semi: false,
  bracketSameLine: false,
  endOfLine: 'auto',
  plugins: [pluginTypescript],
}
export function format(source?: string): Promise<string> {
  if (!source) {
    return Promise.resolve('')
  }
  return prettierFormat(source, formatOptions)
}
