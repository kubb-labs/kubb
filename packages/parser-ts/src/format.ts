import { format as prettierFormat } from 'prettier'
import pluginTypescript from 'prettier/plugins/typescript'

import type { Options } from 'prettier'

const formatOptions: Options = {
  tabWidth: 2,
  printWidth: 160,
  parser: 'typescript',
  singleQuote: true,
  semi: false,
  bracketSameLine: false,
  endOfLine: 'auto',
  plugins: [pluginTypescript],
}

/**
 * Format the generated code based on Prettier
 */
export async function format(source: string) {
  // do some basic linting with the ts compiler
  if (!source) {
    return ''
  }

  return prettierFormat(source, formatOptions)
}
