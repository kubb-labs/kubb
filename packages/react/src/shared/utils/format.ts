import { print } from '@kubb/parser'

import type { Language } from '../../components/Editor.tsx'
import type { DOMElement } from '../../types.ts'

export function format(text: string, node: DOMElement): string {
  const attributes = node.attributes as React.ComponentProps<(typeof Language)>
  try {
    if (attributes.value === 'typescript') {
      return print([], { source: text, noEmitHelpers: false })
    }

    return text
  } catch (e) {
    console.log(e)
  }

  return text
}
