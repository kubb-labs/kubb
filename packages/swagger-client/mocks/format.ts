import { format as prettierFormat } from 'prettier'
import pluginTypescript from 'prettier/plugins/typescript'

import type { Options } from 'prettier'

const formatOptions: Options = {
  tabWidth: 2,
  printWidth: 160,
  parser: 'typescript',
  singleQuote: true,
  semi: false,
  bracketSameLine: false,
  endOfLine: 'auto',
  plugins: [pluginTypescript],
}
export function format(source?: string) {
  if (!source) {
    return ''
  }
  return prettierFormat(source, formatOptions)
}
