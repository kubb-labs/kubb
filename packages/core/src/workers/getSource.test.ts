import path from 'node:path'

import { format } from '../../mocks/format.ts'
import getSource from './getSource.ts'

describe('FileManager utils', () => {
  test('if getFileSource is returning code with imports', async () => {
    const code = getSource({
      file: {
        baseName: 'test.ts',
        path: 'models/ts/test.ts',
        source: 'export type Pet = Pets;',
        imports: [
          {
            name: ['Pets'],
            path: './Pets',
            isTypeOnly: true,
          },
        ],
      },
    })
    const codeWithDefaultImport = getSource({
      file: {
        baseName: 'test.ts',
        path: 'models/ts/test.ts',
        source: 'export type Pet = Pets | Cat; const test = [client, React];',
        imports: [
          {
            name: 'client',
            path: './Pets',
          },
          {
            name: ['Pets', 'Cat'],
            path: './Pets',
            isTypeOnly: true,
          },
          {
            name: 'React',
            path: './React',
          },
        ],
      },
    })
    const codeWithDefaultImportOrder = getSource({
      file: {
        baseName: 'test.ts',
        path: 'models/ts/test.ts',
        source: 'export type Pet = Pets | Cat;\nconst test = [client, React];',
        imports: [
          {
            name: ['Pets', 'Cat'],
            path: './Pets',
            isTypeOnly: true,
          },
          {
            name: 'client',
            path: './Pets',
          },
          {
            name: 'React',
            path: './React',
          },
          {
            name: ['Pets', 'Cat'],
            path: './Pets',
            isTypeOnly: true,
          },
        ],
      },
    })

    expect(await format(code)).toMatchSnapshot()
    expect(await format(codeWithDefaultImport)).toMatchSnapshot()
    expect(await format(codeWithDefaultImportOrder)).toMatchSnapshot()
  })

  test('if getFileSource is returning code with imports and default import', async () => {
    const code = getSource({
      file: {
        baseName: 'test.ts',
        path: 'models/ts/test.ts',
        source: 'export type Pet = Pets;',
        imports: [
          {
            name: 'Pets',
            path: './Pets',
            isTypeOnly: true,
          },
        ],
      },
    })
    expect(await format(code)).toMatchSnapshot()
  })

  test('if getFileSource is returning code with exports and exports as', async () => {
    const codeImport = getSource({
      file: {
        path: path.resolve('./src/models/file1.ts'),
        baseName: 'file1.ts',
        source: `export const test = 2;
      type Test = Pets | Lily | Dog;`,
        imports: [
          {
            name: ['Pets'],
            path: './Pets',
            isTypeOnly: true,
          },
          {
            name: ['Lily'],
            path: './Pets',
            isTypeOnly: true,
          },
          {
            name: 'Dog',
            path: './Dog',
            isTypeOnly: true,
          },
        ],
      },
    })

    const codeExport = getSource({
      file: {
        path: path.resolve('./src/models/file1.ts'),
        baseName: 'file1.ts',
        source: '',
        exports: [
          {
            name: ['Pets'],
            path: './Pets',
            isTypeOnly: true,
          },
          {
            name: ['Lily'],
            path: './Pets',
            isTypeOnly: true,
          },
          {
            name: 'Dog',
            asAlias: true,
            path: './Dog',
            isTypeOnly: true,
          },
        ],
      },
    })

    expect(await format(codeImport)).toMatchSnapshot()
    expect(await format(codeExport)).toMatchSnapshot()
  })

  test('if getFileSource is setting `process.env` based on `env` object', async () => {
    const codeImport = getSource({
      file: {
        path: path.resolve('./src/models/file1.ts'),
        baseName: 'file1.ts',
        source: 'export const hello = process.env.HELLO;',
        imports: [
          {
            name: ['Pets'],
            path: './Pets',
            isTypeOnly: true,
          },
        ],
        env: {
          HELLO: `"world"`,
        },
      },
    })

    const codeImportAdvanced = getSource({
      file: {
        path: path.resolve('./src/models/file1.ts'),
        baseName: 'file1.ts',
        source: 'export const hello = process.env["HELLO"]; type Test = Pets;',
        imports: [
          {
            name: ['Pets'],
            path: './Pets',
            isTypeOnly: true,
          },
        ],
        env: {
          HELLO: `"world"`,
        },
      },
    })

    const codeImportDeclareModule = getSource({
      file: {
        path: path.resolve('./src/models/file1.ts'),
        baseName: 'file1.ts',
        source: `
      declare const TEST: string;

      export const hello = typeof TEST !== 'undefined' ? TEST : undefined
      type Test = Pets;
      `,
        imports: [
          {
            name: ['Pets'],
            path: './Pets',
            isTypeOnly: true,
          },
        ],
        env: {
          TEST: `"world"`,
        },
      },
    })

    expect(await format(codeImport)).toMatchSnapshot()
    expect(await format(codeImportAdvanced)).toMatchSnapshot()
    expect(await format(codeImportDeclareModule)).toMatchSnapshot()
  })

  test('if combineExports is filtering out duplicated exports', async () => {
    const code = getSource({
      file: {
        path: path.resolve('./src/models/file1.ts'),
        baseName: 'file1.ts',
        source: '',
        exports: [
          {
            path: './models',
            name: undefined,
            isTypeOnly: true,
          },
          {
            path: './models',
            isTypeOnly: false,
          },
          {
            path: './models',
            isTypeOnly: false,
            asAlias: true,
            name: 'test',
          },
        ],
      },
    })

    expect(await format(code)).toMatchSnapshot()
  })

  test('if combineImports is filtering out duplicated imports', async () => {
    const code = getSource({
      file: {
        path: path.resolve('./src/models/file1.ts'),
        baseName: 'file1.ts',
        source: 'const test = models; type Test = Config;',
        imports: [
          {
            path: './models',
            name: 'models',
            isTypeOnly: true,
          },
          {
            path: './models',
            name: ['Config'],
            isTypeOnly: true,
          },
          {
            path: './models',
            name: 'models',
            isTypeOnly: false,
          },
        ],
      },
    })

    expect(await format(code)).toMatchSnapshot()

    const codeWithoutSource = getSource({
      file: {
        path: path.resolve('./src/models/file1.ts'),
        baseName: 'file1.ts',
        source: '',
        imports: [
          {
            path: './models',
            name: 'models',
            isTypeOnly: true,
          },
          {
            path: './models',
            name: ['Config'],
            isTypeOnly: true,
          },
          {
            path: './models',
            name: 'models',
            isTypeOnly: false,
          },
        ],
      },
    })

    expect(await format(code)).toMatchSnapshot()
  })

  test.todo('if combineImports is excluding imports when import path and file path are the same')
})
