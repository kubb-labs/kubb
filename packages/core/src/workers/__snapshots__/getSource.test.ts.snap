// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FileManager utils > if combineExports is filtering out duplicated exports 1`] = `
"export * as test from './models'
export type * from './models'
"
`;

exports[`FileManager utils > if combineImports is filtering out duplicated imports 1`] = `
"import type models from './models'
import type { Config } from './models'

const test = models
type Test = Config
"
`;

exports[`FileManager utils > if combineImports is filtering out duplicated imports 2`] = `
"import type models from './models'
import type { Config } from './models'

const test = models
type Test = Config
"
`;

exports[`FileManager utils > if getFileSource is returning code with exports and exports as 1`] = `
"import type { Pets, Lily } from './Pets'
import type Dog from './Dog'

export const test = 2
type Test = Pets | Lily | Dog
"
`;

exports[`FileManager utils > if getFileSource is returning code with exports and exports as 2`] = `
"export type { Pets, Lily } from './Pets'
export type * as Dog from './Dog'
"
`;

exports[`FileManager utils > if getFileSource is returning code with imports 1`] = `
"import type { Pets } from './Pets'

export type Pet = Pets
"
`;

exports[`FileManager utils > if getFileSource is returning code with imports 2`] = `
"import client from './Pets'
import React from './React'
import type { Pets, Cat } from './Pets'

export type Pet = Pets | Cat
const test = [client, React]
"
`;

exports[`FileManager utils > if getFileSource is returning code with imports 3`] = `
"import client from './Pets'
import React from './React'
import type { Pets, Cat } from './Pets'

export type Pet = Pets | Cat
const test = [client, React]
"
`;

exports[`FileManager utils > if getFileSource is returning code with imports and default import 1`] = `
"import type Pets from './Pets'

export type Pet = Pets
"
`;

exports[`FileManager utils > if getFileSource is setting \`process.env\` based on \`env\` object 1`] = `
"export const hello = process.env.HELLO
"
`;

exports[`FileManager utils > if getFileSource is setting \`process.env\` based on \`env\` object 2`] = `
"import type { Pets } from './Pets'

export const hello = process.env['HELLO']
type Test = Pets
"
`;

exports[`FileManager utils > if getFileSource is setting \`process.env\` based on \`env\` object 3`] = `
"import type { Pets } from './Pets'

declare const TEST: string

export const hello = typeof TEST !== 'undefined' ? TEST : undefined
type Test = Pets
"
`;
